File: Core/src/main/java/com/hqq/core/ui/BaseViewBuilderHolder.java
package com.hqq.core.ui;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;

import com.hqq.core.lifecycle.BaseLifecycleObserver;
import com.hqq.core.ui.builder.ICreateRootView;

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.core.ui
 * @FileName :   BaseViewBuilderHolder
 * @Date : 2019/10/31 0031  下午 1:23
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public abstract class BaseViewBuilderHolder extends ViewHolder implements ICreateRootView.IBaseViewBuilderHolder, BaseLifecycleObserver {

    private ViewGroup mParentView;
    Activity mActivity;

    /**
     * 构建activity
     *
     * @param appCompatActivity
     * @param parentView
     */
    public BaseViewBuilderHolder builder(AppCompatActivity appCompatActivity, ViewGroup parentView) {
        createRootView(parentView, appCompatActivity, appCompatActivity, appCompatActivity.getLifecycle());
        return this;
    }

    /**
     * 转移fragment
     *
     * @param fragment
     * @param parentView
     */
    public BaseViewBuilderHolder builder(Fragment fragment, ViewGroup parentView) {
        createRootView(parentView, fragment.getActivity(), fragment.getContext(), fragment.getLifecycle());
        return this;
    }

    @Override
    public void createRootView(ViewGroup parentView, Activity activity, Context context, Lifecycle lifecycle) {
        mParentView = parentView;
        mActivity = activity;
        if (getLayoutViewId() <= 0) {
            mConvertView = getLayoutView(mParentView);
        } else {
            mConvertView = LayoutInflater.from(context).inflate(getLayoutViewId(), mParentView, false);
        }
        lifecycle.addObserver(this);
        initView();
    }


    @Override
    public void initDefConfig() {

    }

    @Override
    public void onClick(View view) {

    }

    @Override
    public View getLayoutView(ViewGroup viewGroup) {
        return null;
    }

    @Override
    public void onCrete() {

    }

    @Override
    public void onResume() {

    }

    @Override
    public void onStop() {

    }

    @Override
    public void onPause() {

    }

    @Override
    public void onStart() {

    }

    @Override
    public void onAny() {

    }

    @Override
    public void onDestroy() {
        removeFromParent();
    }

    public void addToParent() {
        if (mParentView != null && mConvertView != null) {
            mParentView.addView(mConvertView);
        }
    }

    public void removeFromParent() {
        ViewParent parent = mConvertView.getParent();
        if (parent != null) {
            ((ViewGroup) parent).removeView(mConvertView);
        }
    }

}

File: Core/src/main/java/com/hqq/core/ui/BaseViewBuilderHolder.kt
package com.hqq.core.ui

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import com.hqq.core.lifecycle.BaseLifecycleObserver
import com.hqq.core.ui.base.IRootView.IBaseViewBuilderHolder
import java.lang.ref.WeakReference

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.core.ui
 * @FileName :   BaseViewBuilderHolder
 * @Date : 2019/10/31 0031  下午 1:23
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
abstract class BaseViewBuilderHolder : ViewHolder(), IBaseViewBuilderHolder, BaseLifecycleObserver {
    private var parentView: ViewGroup? = null

    var activity: WeakReference<Activity>? = null

    fun getActivity(): Activity? {
        return activity?.get()
    }

    /**
     * 构建activity
     *
     * @param appCompatActivity
     * @param parentView
     */
    fun builder(appCompatActivity: AppCompatActivity, parentView: ViewGroup?): BaseViewBuilderHolder {
        createRootView(parentView, appCompatActivity, appCompatActivity, appCompatActivity.lifecycle)
        return this
    }

    /**
     * 转移fragment
     *
     * @param fragment
     * @param parentView
     */
    fun builder(fragment: Fragment, parentView: ViewGroup?): BaseViewBuilderHolder {
        createRootView(parentView, fragment.activity, fragment.context, fragment.lifecycle)
        return this
    }

    override fun createRootView(parentView: ViewGroup?, activity: Activity?, context: Context?, lifecycle: Lifecycle) {
        this.parentView = parentView
        this.activity = WeakReference<Activity>(activity)
        convertView = if (layoutViewId <= 0) {
            getLayoutView(this.parentView!!)
        } else {
            LayoutInflater.from(context).inflate(layoutViewId, this.parentView, false)
        }
        lifecycle.addObserver(this)
        initView()
    }

    override fun initConfig() {}
    override fun onClick(view: View) {}
    override fun getLayoutView(viewGroup: ViewGroup): View? {
        return null
    }

    override fun onCrete() {}
    override fun onResume() {}
    override fun onStop() {}
    override fun onPause() {}
    override fun onStart() {}
    override fun onAny() {}
    override fun onDestroy() {
        removeFromParent()
    }

    fun addToParent() {
        if (parentView != null && convertView != null) {
            parentView!!.addView(convertView)
        }
    }

    fun removeFromParent() {
        val parent = convertView!!.parent
        if (parent != null) {
            (parent as ViewGroup).removeView(convertView)
        }
    }


}
